lightning_client.go

```diff
@@ -2703,3 +2829,92 @@ func (s *lightningClient) NetworkInfo(ctx context.Context) (*NetworkInfo,
 		NumZombieChans:       resp.NumZombieChans,
 	}, nil
 }
+
+// AcceptorFunction is the signature used for functions passed to our channel
+// acceptor.
+type AcceptorFunction func(context.Context,
+	*AcceptorRequest) (*AcceptorResponse, error)
+
+// AcceptChannel create a channel acceptor using the accept function passed in.
+// The timeout provided will be used to timeout the passed accept closure when
+// it exceeds the amount of time we allow. Note that this amount should be
+// strictly less than lnd's chanacceptor timeout parameter.
+func (s *lightningClient) AcceptChannel(ctx context.Context,
+	timeout time.Duration, accept AcceptorFunction) (chan error, error) {
+
+	acceptStream, err := s.client.ChannelAcceptor(
+		s.adminMac.WithMacaroonAuth(ctx),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	errChan := make(chan error, 1)
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		for {
+			select {
+			case <-ctx.Done():
+				errChan <- ctx.Err()
+				return
+
+			default:
+			}
+
+			request, err := acceptStream.Recv()
+			if err != nil {
+				errChan <- fmt.Errorf("channel acceptor "+
+					"receive failed: %v", err)
+
+				return
+			}
+
+			accReq, err := newAcceptorRequest(request)
+			if err != nil {
+				errChan <- fmt.Errorf("invalid request sent "+
+					"from lnd: %v", err)
+
+				return
+			}
+
+			// Create a child context for our accept function which
+			// will timeout after the timeout period provided.
+			ctxt, cancel := context.WithTimeout(ctx, timeout)
+
+			resp, err := accept(ctxt, accReq)
+			if err != nil {
```

Comment:
Could cancel directly before this line as nothing async is going on after this anymore (or at least shouldn't). Then we could get rid of the other three calls.